test/cocotb/osdtestlib/debug_interconnect.py

```diff
@@ -33,23 +59,97 @@ def transmit_flit(self, dut, data, last):
-                 payload):
+    def __init__(self, destination=1, source=0,
+                 packet_type=PACKET_TYPE_REG,
+                 packet_subtype=PACKET_SUBTYPE_REQ_READ_REG_16,
+                 payload=[]):
 
-        self.destination    = destination
-        self.source         = source
-        self.packet_type    = packet_type
+        self.new = 1
+
+        self.destination = destination
+        self.source = source
+        self.packet_type = packet_type
         self.packet_subtype = packet_subtype
-        self.payload        = payload
+        self.payload = payload
+
+    def from_raw1(self, flits):
+
+        # Check the 1st flit
+        if (flits[0] != 0x0000):
+
+            raise DebugEventFailedException(message="First Flit should be 0, \
+            got %d" % flits[0])
+
+        # Check the 2nd flit
+        elif ((((flits[1] & 0xF000) >> 12) != 0x0008)
+              | ((flits[1] & 0x0400) != 0x0000)):
+
+            raise DebugEventFailure(message="1000x0xxxxxxxxxx is what the 2nd\
+             flit should look like\n%b is what was returned" % flits[1])
+
+        # Status info: trace overflow!
+        elif (flits[1] & 0x0800):
+
+            # This is how the uppermost 6 bits are supposed to be
+            if (((flits[2] & 0xFC00) >> 10) != 0x20):
+                raise DebugEventFailure(message="100000xxxxxxxxxx is what the \
+                3rd flit should look like\n%b is what was returned" % flits[2])
+
+            # In this case, the entire packet should consist of only 3 Flits!
+            elif (len(flits) != 3):
+
+                raise DebugEventFailure(message="%d instead of 3 Flits were\
+                 returned" % len(flits))
+
+            else:
+                self.source = flits[1] & 0x03FF
+                self.packet_type = 0x0002
+                self.packet_subtype = len(flits)
+
+                # The payload is contained in the lowest 10 bits
+                self.payload = (flits[2] & 0x03FF)
+
+                # Display it in hexadezimal
+                dut._log.info("Trace overflow occurred!")
+                dut._log.info("Trace data:")
+                dut._log.info('[{}]'.format(', '.join(hex(x) for x in
+                                                      self.payload)))
+
+        else:
+
+            # Evaluation of the 3rd and 4th flit is obsolete since their
+            # content is determined by the timestamp value during the
+            # generation of the trace event
+
+            self.source = flits[1] & 0x03FF
+            self.packet_type = 0x0002
+            self.packet_subtype = len(flits)
+            self.payload = flits[2:len(flits)]
+
+        self.new = 0
+
+
+    def from_raw2(self, flits):
+
+        self.destination = flits[0]
+        self.source = flits[1]
+        self.packet_type = ((flits[2] & 0xC000) >> 14)
+        self.packet_subtype = ((flits[2] & 0x3C00) >> 10)
+        self.payload = flits[3:len(flits)]
+
+        self.new = 1
```

Comment:
A couple notes on the constructor and the two from_raw* methods:
* What you want are two ways to construct a DiPacket: one constructor creating a DiPacket out of the individual fields, the other one creating a DiPacket from an array of flits.
* Having two constructors is not possible in Python. There are a couple way around that. I'd propose going for the following approach:
  * Make the actual constructor (__init__) create an empty packet, i.e. it takes no arguments at all (other than self, of course).
  * Create two functions to populate the data fields in one go: one function `set_flits(self, flits)` which takes raw data from the NoC and populates the internal fields with it. That's what you have in the from_raw functions right now. The nice thing about this function is that it's a symmetric version to get_flits(). [Better Python style would be to mark these functions property/setter, but that's for later.]
  * Another function which sets source, destination, etc. in one go, called (for example) set_contents(destination, source, ...)
* In your from_raw1 function, you added code beyond the parsing of the packet into fields. Note that DiPacket represents any arbitrary packet. You can only check for obvious violations of the packet structure, nothing more. For example, don't try to enforce a particular source or destination, or try to understand the contents of the packet beyond the packet structure.